---
title: "FindIntronicGenes"
author: "Emanuela Damieto"
date: "2022-09-20"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
if (!require("BiocManager", quietly = TRUE))
  install.packages("BiocManager")

BiocManager::install("GenomicRanges")
```

## Import data and inspect data


```{r import annotation file}
#Set the directory and import the data
library("here")
spruce_pine <- read.table(gzfile(here("reference/gff3/Picab02_codingAll.gff3.gz")), sep = "\t")
#gff = general feature format

#Add the header
header <- c("seqid", "source", "type", "start", "end", "score", "strand", "phase", "attributes")
colnames(spruce_pine) <- header


#Subset the dataframe to retrieve the information that we want
genes <- subset(spruce_pine, type=="gene")
introns <- subset(spruce_pine, type=="intron")  
exons <- subset(spruce_pine, type=="exon")
cds <- subset(spruce_pine, type=="CDS")
mrna <- subset(spruce_pine, type=="mRNA")

```

## Introns in transcripts
```{r import annotation file}
#too many introns in findOverlap object -> do it manually from introns object
parent <- sub(".*=","",introns$attributes)
table_parents <- table(parent)
head(table(parent))
max(table_parents) #74
table_parents[which.max(table_parents)]
mrna[str_detect(mrna$attributes, "PICABM00001871025"),]

hist(table_parents, xlab = "# of introns", main="Number of introns in transcripts", xlim=c(0,70), ylim=c(0,90000))
introns$parent <- parent
introns$length <- introns$end-introns$start+1

#do plot but with transcripts instead of genes

ID <- sub(";.*","",mrna$attributes)
ID <- sub("ID=","",ID)
mrna$ID <- ID


introns$counts <- 1

library(dplyr)
#create a new dataframe, group by parent
df_introns_in_genes <- introns %>% group_by(parent)

#summarize
df_introns_in_genes <- df_introns_in_genes %>% summarise(
  counts= sum(counts),
  seqid=unique(seqid),
  length=sum(length),
  parent=unique(parent)
)

df_introns_in_genes[which.max(df_introns_in_genes$counts),]
df_introns_in_genes[which.max(df_introns_in_genes$length),]
introns[introns$parent=="PICABM00006295253",]


df_introns_in_genes_sort <- df_introns_in_genes %>% arrange(desc(length))
head(df_introns_in_genes_sort$parent,9)
strange_mRNA <- head(df_introns_in_genes_sort$parent,9)
introns[introns$parent %in% strange_mRNA,]


ggplot(df_introns_in_genes, aes(x=counts, y=length, text=parent))+
  geom_point()+
  labs(x="# introns", y="Intron length", title="Scatterplot of intron length and occurrence")


boxplot(log10(df_introns_in_genes$length), main= "Boxplot of intron length and occurrence")


df_introns_in_genes_without_outliers <- df_introns_in_genes %>% filter(!parent %in% strange_mRNA,)

ggplot(df_introns_in_genes_without_outliers, aes(x=counts, y=length, text=parent))+
  geom_point()+
  labs(x="# introns", y="Intron length", title="Scatterplot of intron length and occurrence")

comparisonplot(df_introns_in_genes_without_outliers$counts, df_introns_in_genes_without_outliers$length, xlab ="# introns", ylab = "Intron length", main="Comparison plot of number of introns vs intron length", xlim=c(0,100))

comparisonplot(log10(df_introns_in_genes_without_outliers$counts), log10(df_introns_in_genes_without_outliers$length), xlab ="log # introns", ylab = "log intron length", main="Comparison plot of number of introns vs intron length", xlim= c(0,5))

ggplot(df_introns_in_genes_without_outliers, aes(x=counts, y=length))+
  geom_density_2d()+
  labs(x="# introns", y="Intron length", title="Density plot of intron length and occurrence")

hist(df_introns_in_genes_without_outliers$length, xlim=c(0,700000), main="Histogram of intron length", xlab="Intron length")

```

## Analysis of the two peaks

```{r}
#set the threshold of log intron length to 4
long_introns <- df_introns_in_genes_without_outliers %>% filter(log10(length)>=4)
short_introns <- df_introns_in_genes_without_outliers %>% filter(log10(length)<4)
long_introns_id <- long_introns$parent
short_introns_id <- short_introns$parent



violin_long <- ggplot(long_introns, aes(log10(counts), log10(length)))+
  geom_violin(fill="light blue")+
  labs(x="log 10 # introns", y="log 10 length of introns", title="Violin plot of long introns")

violin_small <- ggplot(short_introns, aes(log10(counts), log10(length)))+
  geom_violin(fill="light blue")+
  labs(x="log 10 # introns", y="log 10 length of introns", title="Violin plot of short introns")

library(ggtree)

multiplot(violin_long, violin_small, ncol=2)



#comparison intron length vs transcript length 
mrna_length <- mrna$end-mrna$start+1
mrna$mrna_length <- mrna_length


mrna_id_length <- mrna %>% select(ID, mrna_length)
#long introns
long_introns_mrna <- long_introns %>% left_join(mrna_id_length, by= c("parent"="ID"), suffix=c("intron", "mRNA"))


ggplot(long_introns_mrna, aes(x=length, y=mrna_length))+
  geom_point()+
  labs(x="intron introns", y="transcript length", title="Scatterplot of intron length and transcript length")

intron_percentage <- long_introns_mrna$length/long_introns_mrna$mrna_length
long_introns_mrna <- long_introns_mrna %>% add_column(intron_percentage)

hist(long_introns_mrna$intron_percentage, breaks= 20, xlim=c(0.7,1), ylim=c(0,50000), xlab="intron length/transcript length", main="Histogram of intron length/transcript length in long introns")
#short introns
short_introns_mrna <- short_introns %>% left_join(mrna_id_length, by= c("parent"="ID"), suffix=c("intron", "mRNA"))


ggplot(short_introns_mrna, aes(x=length, y=mrna_length))+
  geom_point()+
  labs(x="intron introns", y="transcript length", title="Scatterplot of intron length and transcript length")

intron_percentage_short <- short_introns_mrna$length/short_introns_mrna$mrna_length
short_introns_mrna <- short_introns_mrna %>% add_column(intron_percentage_short)

hist(short_introns_mrna$intron_percentage_short, breaks= 20, xlab="intron length/transcript length", main="Histogram of intron length/transcript length in short introns")

#average intron length and sd
mean(long_introns_mrna$length) #164546.3
sd(long_introns_mrna$length) #131363.6

mean(short_introns_mrna$length) #1759.762
sd(short_introns_mrna$length) #2133.828



boxplot(list(long_introns=log10(long_introns_mrna$length), short_introns=log10(short_introns_mrna$length)), ylab="log intron length", main="Boxplot of cumulative intron length", col="bisque")


```
## Intron length vs exon length 
```{r}
#create a dataset with exon length and exon counts 
parent_exon <- sub(".*=","",exons$attributes)
table_parents_exon <- table(parent_exon)
head(table(parent_exon))
max(table_parents_exon) #75
table_parents[which.max(table_parents)]
mrna[str_detect(mrna$attributes, "PICABM00001871025"),]

```


## Intron length vs CDS length 

```{r}
head(cds)
parent_cds <- sub(".*=","",cds$attributes)
table_parents_cds <- table(parent_cds)
head(table(parent_cds))
table_parents_cds <- table_parents_cds[-which(table_parents_cds=="13499")]
max(table_parents_cds) #65
table_parents_cds[which.max(table_parents_cds)] #PICABM00011984427
mrna[str_detect(mrna$attributes, "PICABM00011984427"),]



```


## Average intron length 

```{r}
#average intron length per gene (and sd -> do a boxplot)
#create a new dataframe, group by parent
df_introns_in_genes <- introns %>% group_by(parent)
df_introns_in_genes_mean <- df_introns_in_genes %>% summarise(
  counts= sum(counts),
  seqid=unique(seqid),
  length=mean(length),
  parent=unique(parent)
)

df_introns_in_genes_mean_without_outliers <- df_introns_in_genes_mean %>% filter(!parent %in% strange_mRNA,)

head(df_introns_in_genes_mean_without_outliers)

comparisonplot(df_introns_in_genes_mean_without_outliers$counts, df_introns_in_genes_mean_without_outliers$length, xlab ="# introns", ylab = "Intron length", main="Comparison plot of number of introns vs intron length", xlim=c(0,100))

comparisonplot(log10(df_introns_in_genes_mean_without_outliers$counts), log10(df_introns_in_genes_mean_without_outliers$length), xlab ="log # introns", ylab = "log intron length", main="Comparison plot of number of introns vs intron length", xlim= c(0,5))


long_introns_mean <- df_introns_in_genes_mean_without_outliers %>% filter(log10(length)>3.5)
short_introns_mean <- df_introns_in_genes_mean_without_outliers %>% filter(log10(length)<3.5)



boxplot(list(long_introns=log10(long_introns_mean$length),short_introns=log10(short_introns_mean$length)), ylab="log intron length", main="Boxplot of average intron length in long and short introns", col="bisque")

```
## Analysis of intron length in small and long transcript
```{r}

df_introns_in_genes_without_outliers_no_summarise <- introns %>% filter(!parent %in% strange_mRNA,)

df_long_introns <- df_introns_in_genes_without_outliers_no_summarise %>% filter(parent %in% long_introns_id,)

df_short_introns <- df_introns_in_genes_without_outliers_no_summarise %>% filter(parent %in% short_introns_id,)

boxplot(list(long_introns=log10(df_long_introns$length), short_introns=log10(df_short_introns$length)), ylab="log intron length",main="Boxplot of intron length", notch=TRUE)

```
## Weighted intron length 
```{r}
#measure introns length as the ratio between the length of the longest intron and the sum of intron length (length longest intorn/sum of intron length)
df_introns_in_genes <- introns %>% group_by(parent)
df_introns_in_genes_weight <- df_introns_in_genes %>% summarise(
  counts= sum(counts),
  seqid=unique(seqid),
  length=max(length)/sum(length),
  parent=unique(parent)
)

df_introns_in_genes_weight_without_outliers <- df_introns_in_genes_weight %>% filter(!parent %in% strange_mRNA,)

head(df_introns_in_genes_weight_without_outliers)

#PICABM00000054819: 197/297=0.66
#PICABM00000054842: 1
#PICABM00000054850: 222/318=0.698
df_introns_in_genes_weight_without_outliers[which(df_introns_in_genes_weight_without_outliers$parent=="PICABM00000054850"),]

#comparison plot of the weighted intron length 
comparisonplot(df_introns_in_genes_weight_without_outliers$counts, df_introns_in_genes_weight_without_outliers$length, xlab ="# introns", ylab = "Weighted intron length", main="Comparison plot of number of introns vs intron length", ylim=c(0,1))


long_introns_id <- long_introns$parent
short_introns_id <- short_introns$parent

df_long_introns_weight <- df_introns_in_genes_weight_without_outliers %>% filter(parent %in% long_introns_id,)
df_short_introns_weight <- df_introns_in_genes_weight_without_outliers %>% filter(parent %in% short_introns_id,)

df_long_introns_mean <- df_introns_in_genes_mean_without_outliers %>% filter(parent %in% long_introns_id,)
df_short_introns_mean <- df_introns_in_genes_mean_without_outliers %>% filter(parent %in% short_introns_id,)

df_long_introns_sum <- df_introns_in_genes_without_outliers %>% filter(parent %in% long_introns_id,)
df_short_introns_sum <- df_introns_in_genes_without_outliers %>% filter(parent %in% short_introns_id,)

boxplot(list(long_introns=df_long_introns_weight$length, short_introns=df_short_introns_weight$length), ylab="Weighted intron length",main="Boxplot of weighted intron length", notch=TRUE, col="bisque")

#comparison plot of long transcripts 
comparisonplot(df_long_introns_weight$counts, df_long_introns_weight$length, xlab ="# introns", ylab = "Weighted intron length", main="Comparison plot of number of introns vs intron length in long transcripts", ylim=c(0,1))

#comparison plot of short transcripts 
comparisonplot(df_short_introns_weight$counts, df_short_introns_weight$length, xlab ="# introns", ylab = "Weighted intron length", main="Comparison plot of number of introns vs intron length in short transcripts", ylim=c(0,1))


#Histogram of # of introns
hist(df_introns_in_genes_weight_without_outliers$counts, main="Histogram of the number of introns per transcript", xlab="# introns", xlim=c(0,80))


#Do 1-15 with step=1 and 15+ 
df_introns_in_genes_weight_without_outliers%>% group_by(counts) %>% boxplot()


ggplot(df_introns_in_genes_weight_without_outliers)+
  geom_boxplot(aes(x=factor(counts ), y=length))+
  labs(title="Boxplot of the weighted intron length", x="# introns", y="Weighted intron length")

ggplot(df_introns_in_genes_weight_without_outliers %>% mutate(counts=ifelse(counts>15,15,counts)))+
  geom_boxplot(aes(x=factor(counts ), y=length))+
  labs(title="Boxplot of the weighted intron length", x="# introns", y="Weighted intron length")



#Boxplot of long transcripts
ggplot(df_long_introns_weight  %>% mutate(counts=ifelse(counts>15,15,counts)))+
  geom_boxplot(aes(x=factor(counts), y=length), notch=TRUE)+
  labs(title="Boxplot of the weighted intron length in long transcripts", x="# introns", y="Weighted intron length")

median_long_introns <- df_long_introns_weight  %>% mutate(counts=ifelse(counts>15,15,counts)) %>% group_by(counts) %>% summarise(length= median(length))
median_long_introns$type <- "long"

#Boxplot of short transcripts
ggplot(df_short_introns_weight %>% mutate(counts=ifelse(counts>15,15,counts)))+
  geom_boxplot(aes(x=factor(counts), y=length), notch=TRUE)+
  labs(title="Boxplot of the weighted intron length in short transcripts", x="# introns", y="Weighted intron length")

median_short_introns <- df_short_introns_weight  %>% mutate(counts=ifelse(counts>15,15,counts)) %>% group_by(counts) %>% summarise(length= median(length))
median_short_introns$type <- "short"


median <- union(median_long_introns, median_short_introns)

ggplot(median, aes(counts, length, color=type))+
  geom_point()+
  geom_smooth(formula = y ~ splines::ns(x, 1))+
  labs(title="Plot of the median of long and short trancripts", x="# introns", y="Median weighted length", ylim=c(0,1))


#length vs cumulative length
plot(df_introns_in_genes_without_outliers$length,df_introns_in_genes_weight_without_outliers$length, main="Scatterplot of cumulative length vs weighted length", xlab="Cumulative length", ylab="Weighted length") 
  #abline(reg = lm(df_introns_in_genes_weight_without_outliers$length~df_introns_in_genes_without_outliers$length), col="red")

comparisonplot(df_introns_in_genes_without_outliers$length,df_introns_in_genes_weight_without_outliers$length, main="Comparisonplot of cumulative length vs weighted length", xlab="Cumulative length", ylab="Weighted length", ylim=c(0,1))

#long transcript
plot(df_long_introns_sum$length,df_long_introns_weight$length, main="Scatterplot of cumulative length vs weighted length in long transcripts", xlab="Cumulative length", ylab="Weighted length")

comparisonplot(df_long_introns_sum$length,df_long_introns_weight$length, main="Comparisonplot of cumulative length vs weighted length in long transcripts", xlab="Cumulative length", ylab="Weighted length", ylim=c(0,1))

#short transcript
plot(df_short_introns_sum$length,df_short_introns_weight$length, main="Scatterplot of cumulative length vs weighted length in short transcripts", xlab="Cumulative length", ylab="Weighted length") 

comparisonplot(df_short_introns_sum$length,df_short_introns_weight$length, main="Comparisonplot of cumulative length vs weighted length in short transcripts", xlab="Cumulative length", ylab="Weighted length", ylim=c(0,1))

```





## Genomic Ranges package
This package is used to find the introns inside each gene. 
```{r data pre-processing}
#Use genomic ranges package on genes and introns ;)
#Genomic Ranges works with S4 vectors

mrna_clean <- mrna %>% separate(attributes, c("ID", "Parent", "Name"), ";") %>% mutate (ID= sub("ID=","", ID), Parent= sub("Parent=","",Parent), Name= sub("Name=","",Name)) %>% filter(!ID %in% strange_mRNA,)

#specify the columns you want in the granges objects
library("GenomicRanges")
ggenes <- GRanges(genes)
gintrons <- GRanges(introns)


#Create the overlap
#gintrons is the query (from), ggenes is the subject (to)
overlap <- findOverlaps(gintrons, ggenes, type = "within")

head(overlap, 20)
intr_in_genes <- as.data.frame(table(overlap@to))

intr_in_genes <- as.data.frame(t(apply(intr_in_genes,1, as.numeric)))
colnames(intr_in_genes) <- c("Gene", "Num_introns")

max(intr_in_genes$Num_introns)
#39546
which.max(intr_in_genes$Num_introns)
#19191
intr_in_genes[which.max(intr_in_genes$Num_introns),]
#gene 32513 with 39546 introns
ggenes[32513,]
#PA_chr08 24453721-1397619250



counts_overlap <- countOverlaps(gintrons, ggenes, type="within")
counts_overlap[which.max(counts_overlap)]
subsetByOverlaps(gintrons, ggenes, type="within")
```

## Analysis of transcripts 

```{r}
gmRNA <- GRanges(mrna)
overlap_transcript <- findOverlaps(gintrons, gmRNA, type = "within")

head(overlap_transcript, 20)
intr_in_transcript <- as.data.frame(table(overlap_transcript@to))

intr_in_transcript <- as.data.frame(t(apply(intr_in_transcript,1, as.numeric)))
colnames(intr_in_transcript) <- c("mRNA", "Num_introns")

which.max(intr_in_transcript$Num_introns)
#19191
intr_in_transcript[which.max(intr_in_transcript$Num_introns),]
#transcript 75295 with 24188 introns
gmRNA[75295,]
#PA_chr05 169244672-1128437636
intr_in_transcript[which.min(intr_in_transcript$Num_introns),]

ggplot(intr_in_transcript, aes(Num_introns))+
  geom_density()+
  labs(x="Num of introns in transcripts", title="Density plot of the number of introns in transcripts")

source("../UPSCb-common/src/R/percentile.R")
percentile(intr_in_transcript$Num_introns)
percentile(intr_in_transcript$Num_introns, probs=seq(0.9,1,0.001))


```


## Including Plots of the number of introns with respect to the genes

Plots of the number of introns with respect to the genes. 

```{r plot}
#Try some plots
plot(intr_in_genes, type= "l", lwd=2, main="Plot of gene number vs number of introns", xlab="Gene number", ylab="# introns")


par(mar=c(4,4,4,4))
boxplot(log10(intr_in_genes[2]), main="Boxplot of the number of introns", ylab = "Log of the number of Introns",
        xlim = c(0.5, 1.4))

hist(intr_in_genes[,2], main = "Histogram of the number of introns", 
     xlab = "Number of introns", ylim=c(0,25000),xlim = c(0,300), breaks = 1500)
```

Plots of the number of introns with respect to the genes with ggplot package. 
```{r}
library(ggplot2)
ggplot(intr_in_genes, aes(Gene, Num_introns)) +
  geom_point() +
  coord_cartesian(xlim=c(0,55000), ylim=c(0,7000)) +
  labs(x="Genes", y="Number of introns", title="Scatterplot of the number of introns")

ggplot(intr_in_genes, aes(Gene, Num_introns)) +
  geom_point() +
  coord_cartesian(xlim=c(0,55000), ylim=c(2000,7000)) +
  labs(x="Genes", y="Number of introns", title="Scatterplot of the number of introns above 2000")


ggplot(intr_in_genes, aes(Num_introns)) +
  geom_histogram(binwidth = 2, fill="light blue") +
  coord_cartesian(xlim=c(0,300), ylim=c(0,8000)) +
  labs(x="Number of introns", y="Counts", title="Histogram of the number of introns")

counts_intr <- as.data.frame(table(intr_in_genes$Num_introns))

counts_intr <- as.data.frame(t(apply(counts_intr,1, as.numeric)))
colnames(counts_intr) <- c("Num_introns", "Counts")

ggplot(counts_intr, aes(Counts,Num_introns)) +
  geom_point()+
  labs(title="Plot of the counts of the number of introns")

#Look at the first 10 genes with most introns
#sorted
intr_in_genes_sorted <- intr_in_genes[order(intr_in_genes$Num_introns, decreasing=TRUE),] 
hist(intr_in_genes_sorted$Num_introns, main="Histogram of the frequency of the number of introns",
     xlab="# introns", xlim=c(0,500))
plot(intr_in_genes_sorted, main="Scatterplot of the number of introns withing genes")

ggplot(intr_in_genes_sorted, aes(log10(Num_introns))) +
  geom_boxplot() +
  labs(x="Log # introns", title="Boxplot of the number of introns within genes")

```

## Search for genes that are not in the chromosomes
There are genes that we know in which chromosome they are located but not where ex. PA_chr02_sUL006. There are other genes that we don't know where they are located, maybe they are small circular chromosome, e.g PA_cUP0116. In which we are interested in
```{r}
#Select the seqid that are not in the chromosomes so they don't have chr in the seqid
seqid <- genes[1]
g2 <- grepl("PA_chr", seqid$seqid)

not_chr <- seqid[which(!g2),]
plot(table(not_chr), main="Counts of non chromosomic genes", xlab="Non chromosomic genes",
     ylab="Counts", type="h")
not_chr_counts <- as.data.frame(table(not_chr))


not_chr_counts[which.max(not_chr_counts$Freq),]
```

## Look at gene with the highest number of introns 
```{r}
#Interesting transcript: ID=PICABG00010504706
#check how many exons
interesting_gene <- ggenes[32513,]
gexons <- GRanges(exons)
gcds <- GRanges(cds)
over_exons <- findOverlaps(gexons, interesting_gene, type="within")
#46510 exons
over_introns <- findOverlaps(gintrons, interesting_gene, type="within")
#39546 introns 
over_CDS <- findOverlaps(gcds, interesting_gene, type="within")
#42292 CDS
```

##Look at the 10 genes with the highest number of introns
```{r}
high_freqintr <- head(intr_in_genes_sorted, 10)
most_interesting_genes <- ggenes[high_freqintr$Gene,]
most_interesting_genes
```

##Intron length vs number of introns
```{r}
#the length is the difference between the intron end and the intron start +1 (if not there is an intron with length 0)
intron_length <- introns$end-introns$start+1
boxplot(log10(intron_length), main="Boxplot of intron length", ylab="log intron length")
max(intron_length)
#958772783

df <- data.frame(Gene=overlap@to, Introns=overlap@from,Intron_length=gintrons@ranges@width[overlap@from], seqnames=ggenes@seqnames[subjectHits(overlap),], ranges=ggenes@ranges[subjectHits(overlap),])
merged <- merge.data.frame(df, intr_in_genes, by="Gene", all.x = TRUE)


cor(merged$Num_introns, merged$Intron_length)
#0.003396114

library(tidyverse)

#Group By gene
df_grouped <- merged %>% group_by(Gene)

head(df_grouped)


#summarise by sum
df_new <- df_grouped %>% summarise(
  Gene=unique(Gene),
  Intron_length=sum(Intron_length),
  Num_introns= unique(Num_introns),
  seqnames=unique(seqnames),
  ranges.start=unique(ranges.start),
  ranges.end=unique(ranges.end),
  geneID=unique(geneID)
  
)

head(df_new)

write.csv(df_new, file="Introns_within_genes.csv")

```

```{r}
attributes <- mcols(ggenes)$attributes
att <- strsplit(attributes, split=";")
geneID_tot <- lapply(att, `[[`, 1)   

geneID_introns_in_genes <- as.character(geneID_tot[intr_in_genes$Gene])

df_complete <- df_new %>% mutate(geneID=geneID_introns_in_genes)

library("plotly")

scatterplot <- ggplot(df_complete, aes(x=Num_introns, y=Intron_length, text=geneID))+
  geom_point()+
  labs(x="# introns", y="Intron length", title="Scatterplot of intron length and occurrence")

ggplotly(scatterplot)


```




Plot
```{r}
plot(merged$Num_introns, merged$Intron_length, main="Scatterplot of intron length and occurrence",
     xlab="# introns", ylab="Intron length")


ggplot(merged, aes(Num_introns, Intron_length))+
  geom_point()+
  labs(x="# introns", y="Intron length", title="Scatterplot of intron length and occurrence")

#Density plot
ggplot(merged, aes(Num_introns, Intron_length))+
  geom_density_2d()+
  labs(x="# introns", y="Intron length", title="Density plot of intron length and occurrence")

#Density plot
ggplot(df_new, aes(Num_introns, Intron_length))+
  geom_density_2d()+
  labs(x="# introns", y="Intron length", title="Density plot of intron length and occurrence")

ggplot(df_new, aes(Num_introns, Intron_length))+
  geom_point()+
  labs(x="# introns", y="Intron length", title="Scatterplot of intron length and occurrence")


ggplot(df_new, aes(Num_introns, Intron_length))+
  geom_point()+
  geom_smooth(method="lm")+
  labs(x="# introns", y="Intron length", title="Scatterplot of intron length and occurrence (lm)")

#Transform the axis in log scale 
ggplot(df_new, aes(log10(Num_introns), log10(Intron_length)))+
  geom_point()+
  geom_smooth(method="lm")+
  labs(x="log # introns", y="log intron length", title="Scatterplot of intron length and occurrence (lm)")

linear_model <- lm(Intron_length~Num_introns, df_new)
summary(linear_model)
par(mfrow=c(2,2))
plot(linear_model)

linear_model_2 <- lm(log10(Intron_length)~log10(Num_introns), df_new)
summary(linear_model_2)
par(mfrow=c(2,2))
plot(linear_model_2)

ggplot(df_new, aes(Num_introns, Intron_length))+
  geom_point()+
  geom_smooth(method="gam")+
  labs(x="# introns", y="Intron length", title="Scatterplot of intron length and occurrence (gam)")
#generalized additive models with inetgrated smoothness estimation
library(mgcv)
gen_add_mod <- gam(df_new$Intron_length~s(df_new$Num_introns, bs="cs"))
summary(gen_add_mod)
par(mfrow=c(1,1))
plot(gen_add_mod)

cor(df_new$Num_introns, df_new$Intron_length)
# 0.9681607

ggplot(df_new, aes(Num_introns, Intron_length))+
  geom_point()+
  geom_smooth(method="loess")+
  labs(x="# introns", y="Intron length", title="Scatterplot of intron length and occurrence (loess)")

#Loess regression
loess_mod <- loess(Intron_length~Num_introns, df_new)
summary(loess_mod)
plot(loess_mod)
```

Plot... to be continued
```{r}
ggplot(df_new, aes(Num_introns, log10(Intron_length)))+
  geom_point()+
  labs(x="# introns", y="log(Intron length)", title="Scatterplot of intron length and occurrence")

new_model <- lm(log(Intron_length)~Num_introns, df_new)
summary(new_model)


ggplot(df_new, aes(Num_introns, log10(Intron_length)))+
  geom_point()+
  geom_smooth(method="gam")+
  labs(x="# introns", y="log(Intron length)", title="Scatterplot of intron length and occurrence")

ggplot(df_new, aes(x=Num_introns, y=log10(Intron_length)))+
  geom_point()+
  geom_smooth(method="lm", formula= y~log10(x))+
  labs(x="# introns", y="log(Intron length)", title="Scatterplot of intron length and occurrence")

new_model_2 <- lm(log(Intron_length)~log10(Num_introns), df_new)
summary(new_model_2)

ggplot(df_new, aes(log(Num_introns), log10(Intron_length)))+
  geom_point()+
  geom_smooth(method="lm", se=TRUE)+
  labs(x="log # introns", y="log(Intron length)", title="Scatterplot of intron length and occurrence")

  


library(LSD)

comparisonplot(df_new$Num_introns, df_new$Intron_length, xlab = "# introns", ylab = "Intron length", main="Comparison plot of intron length and occurrence", xlim=c(0,3000), ylim=c(0,10000000))

comparisonplot(df_new$Num_introns, df_new$Intron_length, xlab = "# introns", ylab = "Intron length", cor=TRUE,
               xlim=c(0,80), main="Comparison plot of intron length and occurrence,")

comparisonplot(df_new$Num_introns, log10(df_new$Intron_length), xlab = "# introns", ylab = "log Intron length", cor = TRUE, xlim=c(0,80), ylim=c(0,15), main="Comparison plot of intron length and occurrence,")
#zoom in the x axis
comparisonplot(df_new$Num_introns, log(df_new$Intron_length), xlab = "# introns", ylab = "log Intron length", cor = TRUE, xlim=c(0,20), ylim=c(0,15), main="Comparison plot of intron length and occurrence,")

comparisonplot(log10(df_new$Num_introns), log10(df_new$Intron_length), xlab = "log # introns", ylab = "log Intron length", cor = TRUE, main="Comparison plot of intron length and occurrence,")
abline(y=5)

```



#Analysis of genes within introns

```{r}
# ANALYSIS OF GENES WITHIN INTRONS

#Find genes inside introns
#In ovl object the query object is ggenes and the subject one is gintrons
ovl <- findOverlaps(ggenes, gintrons, type="within") #93090 overlap 
#ovl <- findOverlaps(ggenes, gintrons, type="within",ignore.strand=TRUE)
#ovl is not empty just if you ignore strand so this mean that every time the intron and the gene are in the opposite strand
# so opposite direction and this is interesting 

introns_with_genes <- gintrons[subjectHits(ovl),]
genes_within_introns<- ggenes[queryHits(ovl),]

#check if the genes are in the chromosome 
par(mar=c(5,5,2,5))
barplot(sort(table(as.character(seqnames(ggenes[queryHits(ovl),])))),las=2, cex.names = 0.5,
        ylim = c(0,4000), main="Barplot of the number of genes within introns", ylab = "# of genes within introns")


#same genes within different introns???
head(ovl)
#the same genes is counted multiple times within different introns. This is due how the script is written 



#table with gene_id, intron_id, #of occurrence

#401 unique genes, 46 unique genes within introns, 10 unique not chromosomic genes within introns 

#as.character(seqnames(ggenes[queryHits(ovl),]))


#CREA UN DF WITH GENE, INTRON TOTAL LENGTH, CHR
df_within <- data.frame(Gene=ovl@from, Intron=ovl@to, Intron_length=gintrons@ranges@width[ovl@to], chr= as.character(seqnames(ggenes[queryHits(ovl),])))

#Group By gene
df_grouped_within <- df_within %>% group_by(Gene)

#summarise by sum
df_new_within <- df_grouped_within %>% summarise(
  Intron_length=sum(Intron_length),
  chr=unique(chr)
)

#df_new_within_sort <- arrange(df_new_within, chr)

ggplot(df_new_within, aes(chr, Intron_length))+
  geom_boxplot()+
  labs(title="Boxplot of cumulative intron length of the chromosomes", x="Chr", y="Cumulative intron length")+
  theme(axis.text.x = element_text(angle = 90)) 

ggplot(df_new_within, aes(chr, log10(Intron_length)))+
  geom_boxplot()+
  labs(title="Boxplot of cumulative intron length of the chromosomes", x="Chr", y="Cumulative intron length")+
  theme(axis.text.x = element_text(angle = 90)) 

## Consider just the genes within the chromosomes
df_chr_genes <- df_new_within %>% filter(str_detect(chr,  regex(paste0("^PA_chr[0-9]{2}$"))))

ggplot(df_chr_genes, aes(chr, Intron_length))+
  geom_boxplot()+
  labs(title="Boxplot of cumulative intron length of the chromosomes", x="Chr", y="Cumulative intron length")+
  theme(axis.text.x = element_text(angle = 90)) 

#Transform the y axis in log scale 
ggplot(df_chr_genes, aes(chr, log10(Intron_length)))+
  geom_boxplot()+
  labs(title="Boxplot of cumulative intron length in the chromosomes", x="Chr", y="log of cumulative intron length")+
  theme(axis.text.x = element_text(angle = 90)) 

#Add standard deviation (notch)
ggplot(df_chr_genes, aes(chr, log10(Intron_length)))+
  geom_boxplot(notch=TRUE)+
  labs(title="Boxplot of cumulative intron length of the chromosomes", x="Chr", y="Cumulative intron length")+
  theme(axis.text.x = element_text(angle = 90)) 

#Notches are used in box plots to help visually assess whether the medians of distributions differ. 
#If the notches do not overlap, this is evidence that the medians are different.

#violin plot
ggplot(df_chr_genes, aes(chr, log10(Intron_length)))+
  geom_violin(fill= "light blue")+
  labs(title="Violin plot of cumulative intron length of the chromosomes", x="Chr", y="log cumulative intron length")+
  theme(axis.text.x = element_text(angle = 90)) 

## Consider the genes not in the chromosomes
df_not_chr_genes <- df_new_within %>% filter(str_detect(chr,  regex(paste0("PA_chr[0-9]")), negate = TRUE))
ggplot(df_not_chr_genes, aes(chr, Intron_length))+
  geom_boxplot()+
  labs(title="Boxplot of cumulative intron length of not chromosomic genes", x="Chr", y="Cumulative intron length")+
  theme(axis.text.x = element_text(angle = 90)) 

#Transform the y axis in log scale
ggplot(df_not_chr_genes, aes(chr, log10(Intron_length)))+
  geom_boxplot()+
  labs(title="Boxplot of cumulative intron length of not chromosomic genes", x="Chr", y="Cumulative intron length")+
  theme(axis.text.x = element_text(angle = 90))
```
## Analysis of genes not within introns

```{r}
#look at the complementary dataset -> genes not in introns, drop genes without introns
ovl_compl <- findOverlaps(ggenes, gintrons, type="any", ignore.strand= TRUE)
ovl_compl <- ovl_compl[!ovl_compl %in% ovl]

ovl_compl[ovl_compl@to==0,] #0 objects
ovl_compl[ovl_compl@from==0,] #0 objects

#create the complement dataframe
df_compl <- data.frame(Gene=ovl_compl@to, Intron=ovl_compl@from, Intron_length=gintrons@ranges@width[ovl_compl@from], chr= as.character(seqnames(ggenes[queryHits(ovl_compl),])))

#Group By gene
df_grouped_compl <- df_compl %>% group_by(Gene)

#summarise by sum
df_new_compl <- df_grouped_compl %>% summarise(
  Intron_length=sum(Intron_length),
  chr=unique(chr)
)

#barplot of chromosome 
par(mar=c(5,5,4,5))
barplot(sort(table(as.character(seqnames(ggenes[queryHits(ovl_compl),])))),las=2, cex.names = 0.5,
        ylim = c(0,10000), main="Barplot of the number of genes not within introns", ylab = "# of genes within introns")


#plot

## Consider just the genes within the chromosomes
df_chr_genes_compl <- df_new_compl %>% filter(str_detect(chr,  regex(paste0("^PA_chr[0-9]{2}$"))))

ggplot(df_chr_genes_compl, aes(chr, Intron_length))+
  geom_boxplot()+
  labs(title="Boxplot of cumulative intron length of the chromosomes", x="Chr", y="Cumulative intron length")+
  theme(axis.text.x = element_text(angle = 90)) 

#Transform the y axis in log scale 
ggplot(df_chr_genes_compl, aes(chr, log10(Intron_length)))+
  geom_boxplot()+
  labs(title="Boxplot of cumulative intron length of the chromosomes", x="Chr", y="Cumulative intron length")+
  theme(axis.text.x = element_text(angle = 90)) 

#Add standard deviation (notch)
ggplot(df_chr_genes_compl, aes(chr, log10(Intron_length)))+
  geom_boxplot(notch=TRUE)+
  labs(title="Boxplot of cumulative intron length of the chromosomes", x="Chr", y="Cumulative intron length")+
  theme(axis.text.x = element_text(angle = 90)) 

#violin plot
ggplot(df_chr_genes_compl, aes(chr, log10(Intron_length)))+
  geom_violin(fill= "light blue")+
  labs(title="Violin plot of cumulative intron length of the chromosomes", x="Chr", y="Cumulative intron length")+
  theme(axis.text.x = element_text(angle = 90))
```


```{r}
#Look at the size distribution of the introns per gene
#look at the median of it
#summarise by median
df_median <- df_grouped %>% summarise(
  Intron_length=median(Intron_length),
  Num_introns= unique(Num_introns)
)

head(df_median)

ggplot(df_median, aes(Num_introns, Intron_length))+
  geom_point()+
  labs(x="# introns", y="Intron length", title="Scatterplot of intron length and occurrence")



#look at the mad (the sd but for the median-median absolute deviation mad)
```


```{r}
#add seqnames and ranges in the df_new dataset
#take just one intron and probably it will not be the shortest or the longest but something in between but in this way, you'll lose the info
ovl <- ovl[!duplicated(queryHits(ovl)),]

df_within <- data.frame(Gene=ovl@from, Introns=ovl@to,Intron_length=gintrons@ranges@width[ovl@to], seqnames=ggenes@seqnames[queryHits(ovl),], ranges=ggenes@ranges[queryHits(ovl),])
#countOverlaps(ggenes, gintrons)

#Group By gene
df_grouped_within <- df_within %>% group_by(Gene)

#summarise by sum
df_new_within <- df_grouped_within %>% summarise(
  Gene=unique(Gene),
  Intron_length=sum(Intron_length),
  seqnames=unique(seqnames),
  ranges.start=unique(ranges.start),
  ranges.end=unique(ranges.end)
)
head(df_new_within)

#same genes within different introns???
head(ovl)
#the same genes is counted multiple times within different introns. This is due how the script is written 

write.csv(df_new_within, file="Genes_within_introns.csv")


```

#rtracklayer

```{r}
library("rtracklayer")
gff3 <- import.gff3("../reference/gff3/all_merged_no_overlaps_no_dup_features_gt_gff_after_merge_ensembl_renamed_over_30aa_salmon_redundant_mRNA_removed.gff3.gz")

#Length of the transcript/intron/exon/CDS...
width(gff3)

seqnames(gff3)
ranges(gff3)

#remove outliers
#gff3 before cleaning 3727141
gff3 <- gff3[!gff3$ID %in% strange_mRNA,]
gff3 <- gff3[!as.character(gff3$Parent) %in% strange_mRNA,]
#gff3 after cleaning 3726862

#gff3 <- gff3[!parent %in% strange_mRNA,]
#gff3 <- gff3[!ID %in% strange_mRNA,]

gff3[gff3$ID %in% strange_mRNA,]


gff3_gene <- gff3[which(gff3$type=="gene"),]
gff3_intron <- gff3[which(gff3$type=="intron"),]

gff3_overlap <- findOverlaps(gff3_intron, gff3_gene, type = "within")

gff3_mrna <- gff3[which(gff3$type=="mRNA")]

gff3_exon <- gff3[which(gff3$type=="exon")]

table(gff3_overlap@to)
#max 53186
#which.max 1289727
width(gff3_gene)

max(width(gff3_gene))
#1373165530
which.max(width(gff3_gene))
gff3_gene[32513,]

#PICABG00010504706
longest_gene <- subset(gff3, subset(gff3$Parent=="PICABG00010504706"))
longest_gene$type


gff3_intron$counts <- 1

df_introns <- data.frame(seqnames=seqnames(gff3_intron), ranges= ranges(gff3_intron),strand= strand(gff3_intron),type=gff3_intron$type,
                         ID=gff3_intron$ID, Parent=gff3_intron$Parent, counts=gff3_intron$counts)

df_introns_elaborated <- df_introns %>% group_by(Parent.value) %>%
  summarise(width=sum(ranges.width), counts=sum(counts))


df_genes <- data.frame(seqnames=seqnames(gff3_gene), ranges= ranges(gff3_gene),strand= strand(gff3_gene),type=gff3_gene$type,ID=gff3_gene$ID)

gff3_mrna$counts <- 1
df_mrna <- data.frame(seqnames=seqnames(gff3_mrna), ranges= ranges(gff3_mrna),strand= strand(gff3_mrna),type=gff3_mrna$type,
                         ID=gff3_mrna$ID, Parent=gff3_mrna$Parent, counts=gff3_mrna$counts)


df_mrna_grouped <- df_mrna %>% group_by(Parent.value)

df_mrna_summarize <- df_mrna_grouped %>% summarise(width=sum(ranges.width), counts=sum(counts))

max(df_mrna_summarize$counts)
#76
which.max(df_mrna_summarize$counts)
#51133
df_mrna_summarize[51133,]
#PICABG00017828298

df_genes[which(df_genes$ID=="PICABG00017828298"),]
#290874
df_mrna[which(df_mrna$Parent.value=="PICABG00017828298"),]
#PICABM00017828299 -> longest transcript in the gene with the highest number of transcripts, 290871 bp

df_introns_elaborated[which(df_introns_elaborated$Parent.value=="PICABM00017828299"),]
#9 introns, width 289455 -> 1416 bp exons
df_introns[which(df_introns$Parent.value=="PICABM00017828299"),]

```

##Gene and intron length against CDS length 
```{r}
gff3_cds <- gff3[which(gff3$type=="CDS")]
gff3_cds$counts <- 1

df_cds <- data.frame(seqnames=seqnames(gff3_cds), ranges= ranges(gff3_cds),strand= strand(gff3_cds),type=gff3_cds$type,
                         ID=gff3_cds$ID, Parent=gff3_cds$Parent, counts=gff3_cds$counts)

df_cds_grouped <- df_cds %>% group_by(Parent.value)

df_cds_summarize <- df_cds_grouped %>% summarise(width=sum(ranges.width), counts=sum(counts))

merge_cds_intron <- df_cds_summarize %>% full_join(df_introns_elaborated, by="Parent.value", suffix=c(".cds",".intron"), na.matches=c("na","na"))

#analyze merge object

merge_cds_intron[which(is.na(merge_cds_intron$width.intron)),]
#35965
merge_cds_intron[which(is.na(merge_cds_intron$width.cds)),]
#9



#merge_cds_intron[order(merge_cds_intron$width.intron, decreasing = TRUE),]
ggplot(merge_cds_intron, aes(width.cds, width.intron))+
  geom_point()+
  labs(x="CDS length", y="Intron length", title = "Scatterplot of CDS length vs intron length")

ggplot(merge_cds_intron, aes(width.cds, width.intron))+
  geom_point()+
  scale_x_log10()+
  scale_y_log10()+
  labs(x="CDS length", y="Intron length", title = "Scatterplot of CDS length vs intron length in log scale")




comparisonplot(merge_cds_intron$width.cds, merge_cds_intron$width.intron, xlab="CDS length", ylab="Intron length", main="Comparisonplot of CDS length vs intron length", xlim=c(0,16000))

comparisonplot(log10(merge_cds_intron$width.cds), log10(merge_cds_intron$width.intron), xlab="CDS length", ylab="Intron length", main="Comparisonplot of CDS length vs intron length in log scale")

comparisonplot(merge_cds_intron$width.cds, merge_cds_intron$width.intron, log="xy", xlab="CDS length", ylab="Intron length", main="Comparisonplot of CDS length vs intron length in log scale")
```


```{r}
cds_mrna <- df_cds %>% full_join(df_mrna, by=c("Parent.value"="ID"), suffix=c(".cds",".mrna"), na.matches=c("na","na"))

cds_mrna_grouped <- cds_mrna %>% group_by(ID, Parent.value.mrna)

#cds_mrna_summarize <- cds_mrna_grouped %>% summarise(ranges.width.cds=sum(ranges.width.cds), ranges.width.mrna=unique(ranges.width.mrna))

library("ggplot2")

ggplot(cds_mrna_grouped, aes(ranges.width.cds, ranges.width.mrna))+
  geom_point()+
  labs(x="CDS length", y="Transcript length", title = "Scatterplot of CDS length vs transcript length")


ggplot(cds_mrna_grouped, aes(ranges.width.cds, ranges.width.mrna))+
  geom_point()+
  scale_x_log10()+
  scale_y_log10()+
  labs(x="CDS length", y="Transcript length", title = "Scatterplot of CDS length vs transcript length")


comparisonplot(cds_mrna_grouped$ranges.width.cds, cds_mrna_grouped$ranges.width.mrna, xlab="CDS length", ylab="Transcript length", main="Comparisonplot of CDS length vs transcript length", xlim=c(0,10000))


comparisonplot(log10(cds_mrna_grouped$ranges.width.cds), log10(cds_mrna_grouped$ranges.width.mrna), xlab="CDS length", ylab="Transcript length", main="Comparisonplot of CDS length vs transcript length")

comparisonplot(cds_mrna_grouped$ranges.width.cds, cds_mrna_grouped$ranges.width.mrna, log="xy", xlab="CDS length", ylab="Transcript length", main="Comparisonplot <br> of CDS length vs transcript length")

#54 transcripts without CDS out of 1003638

#zoom long CDS
ggplot(cds_mrna_grouped, aes(ranges.width.cds, ranges.width.mrna))+
  geom_point()+
  xlim(7500,10000)+
  ylim(0,50000)+
  labs(x="CDS length", y="Transcript length", title = "Scatterplot of CDS length vs transcript length")

#Play with plot
ggplot(cds_mrna_grouped, aes(ranges.width.cds, ranges.width.mrna))+
  geom_hex()+
  labs(x="CDS length", y="Transcript length", title = "Scatterplot of CDS length vs transcript length")


ggplot(cds_mrna_grouped, aes(ranges.width.cds, ranges.width.mrna))+
  geom_bin2d(bins = 500)+
  labs(x="CDS length", y="Transcript length", title = "Scatterplot of CDS length vs transcript length")

ggplot(cds_mrna_grouped, aes(ranges.width.cds))+
  geom_histogram()+
  xlim(0,2000)+
  labs(x="CDS length", y="Count", title = "Histogram of CDS length")

```

#Split long transcripts and short transcripts
```{r, CDS length in transcript with long and short introns}
#Split the transcript in two based on intron length. In this way we'll lose the transcript without introns 
long_introns <- df_introns_in_genes_without_outliers %>% filter(log10(length)>=4)
short_introns <- df_introns_in_genes_without_outliers %>% filter(log10(length)<4)
long_introns_id <- long_introns$parent
short_introns_id <- short_introns$parent


#cds_long_transcript <- cds_mrna_grouped %>% filter(Parent.value %in% long_introns_id,)
cds_long_transcript <- cds_mrna_grouped[which(cds_mrna_grouped$Parent.value %in% long_introns_id),]

#cds_short_transcript <- cds_mrna_grouped %>% filter(Parent.value %in% short_introns_id,)
cds_short_transcript <- cds_mrna_grouped[which(cds_mrna_grouped$Parent.value %in% short_introns_id),]


#Solve the problem with the title

comparisonplot(cds_long_transcript$ranges.width.cds, cds_long_transcript$ranges.width.mrna, xlab="CDS length", ylab="Transcript length", main="Comparisonplot of CDS length vs transcript length in transcripts with long introns", xlim=c(0,10000))

#Try with \n
comparisonplot(log10(cds_long_transcript$ranges.width.cds), log10(cds_long_transcript$ranges.width.mrna), xlab="CDS length", ylab="Transcript length", main="CDS length vs transcript length in transcripts with long introns in log scale")

comparisonplot(cds_long_transcript$ranges.width.cds, cds_long_transcript$ranges.width.mrna, log="xy", xlab="CDS length", ylab="Transcript length", main="CDS length vs transcript length in transcripts with long introns in log scale")



comparisonplot(cds_short_transcript$ranges.width.cds, cds_short_transcript$ranges.width.mrna, xlab="CDS length", ylab="Transcript length", main="Comparisonplot CDS length vs transcript length in transcripts with short introns", xlim=c(0,10000))

comparisonplot(log10(cds_short_transcript$ranges.width.cds), log10(cds_short_transcript$ranges.width.mrna), xlab="CDS length", ylab="Transcript length", main="CDS length vs transcript length in transcripts with short introns in log scale")

comparisonplot(cds_short_transcript$ranges.width.cds, cds_short_transcript$ranges.width.mrna, log="xy", xlab="CDS length", ylab="Transcript length", main="CDS length vs transcript length in transcripts with short introns in log scale")

```

```{r}
#compare the length of CDS with the number of exons present in a CDS (use findOverlap?????????????????????????????????)
gff3_exon$counts <- 1
df_exons <- data.frame(seqnames=seqnames(gff3_exon), ranges= ranges(gff3_exon),strand= strand(gff3_exon),type=gff3_exon$type,
                         ID=gff3_exon$ID, Parent=gff3_exon$Parent, counts=gff3_exon$counts)




```


```{r}
#Look where the long introns are located (CDS or UTRs) so plot #exons in the mRNA vs CDS length


df_exons_grouped <- df_exons %>% group_by(Parent.value)

df_exons_summarize <- df_exons_grouped %>% summarise(ranges.width=sum(ranges.width), counts=sum(counts))

cds_exons_summarize <- df_cds_summarize %>% left_join(df_exons_summarize, by="Parent.value", suffix=c(".cds",".exon"))

ggplot(cds_exons_summarize, aes(counts.exon, width))+
  geom_point()+
  labs(title="Scatterplot of the number of exons in a transcript vs CDS length", x="# exons", y="CDS length")

comparisonplot(cds_exons_summarize$counts.exon, cds_exons_summarize$width, xlim=c(0,80))
comparisonplot(log10(cds_exons_summarize$counts.exon), log10(cds_exons_summarize$width), main = "Comparisonplot of the number of exons in a transcript vs CDS length", xlab="# exons", ylab = "CDS length")


head(cds_exons_summarize %>% arrange(desc(counts.exon)))


cds_exons_long_transcript <- cds_exons_summarize[which(cds_exons_summarize$Parent.value %in% long_introns_id),]
cds_exons_short_transcript <- cds_exons_summarize[which(cds_exons_summarize$Parent.value %in% short_introns_id),]

head(cds_exons_long_transcript %>% arrange(desc(counts.exon)))
head(cds_exons_short_transcript %>% arrange(desc(counts.exon)))

ggplot(cds_exons_long_transcript, aes(counts.exon, width))+
  geom_point(alpha=0.5)+
  geom_smooth(method="glm")+
  labs(title="Scatterplot of the number of exons in a transcript vs CDS length\nin transcripts with long introns", x="# exons", y="CDS length")

ggplot(cds_exons_long_transcript, aes(counts.exon, width))+
  geom_hex(bins=50)+
  geom_smooth(method="glm")+
  labs(title="Scatterplot of the number of exons in a transcript vs CDS length\nin transcripts with long introns", x="# exons", y="CDS length")

ggplot(cds_exons_long_transcript, aes(counts.exon, width))+
  geom_point(alpha=0.5)+
  geom_smooth(method="gam")+
  labs(title="Scatterplot of the number of exons in a transcript vs CDS length\nin transcripts with long introns", x="# exons", y="CDS length")


comparisonplot(log10(cds_exons_long_transcript$counts.exon), log10(cds_exons_long_transcript$width), main = "Number of exons in a transcript vs CDS length in transcripts with long introns", xlab="# exons", ylab = "CDS length")

comparisonplot(cds_exons_long_transcript$counts.exon, cds_exons_long_transcript$width, log="xy", main = "Number of exons in a transcript vs CDS length in transcripts with long introns", xlab="# exons", ylab = "CDS length")

ggplot(cds_exons_short_transcript, aes(counts.exon, width))+
  geom_point(alpha=0.5)+
  geom_smooth(method="glm")+
  labs(title="Scatterplot of the number of exons in a transcript vs CDS length\nin transcripts with short introns", x="# exons", y="CDS length")

ggplot(cds_exons_short_transcript, aes(counts.exon, width))+
  geom_point(alpha=0.5)+
  geom_smooth(method="gam")+
  labs(title="Scatterplot of the number of exons in a transcript vs CDS length\nin transcripts with short introns", x="# exons", y="CDS length")


comparisonplot(log10(cds_exons_short_transcript$counts.exon), log10(cds_exons_short_transcript$width), main = "Number of exons in a transcript vs CDS length in transcripts with short introns", xlab="# exons", ylab = "CDS length")

comparisonplot(cds_exons_short_transcript$counts.exon, cds_exons_short_transcript$width, log="xy", main = "Number of exons in a transcript vs CDS length in transcripts with short introns", xlab="# exons", ylab = "CDS length")


#calculate ratio # of CDS introns/# of introns for long and short mRNA????????????????????????????????????????

df_cds[which(df_cds$Parent.value=="PICABM00001871025"),]
df_exons[which(df_exons$Parent.value=="PICABM00001871025"),]


```


```{r, # CDS/# exons}
#Plot the number of exons classed as CDS against the total number of exons

#whole dataset
ggplot(cds_exons_summarize, aes(counts.cds/counts.exon))+
  geom_histogram(bins=25, fill="light blue", color="grey")+
  labs(x="# CDS/# exons", title ="Histogram of the ratio number of CDS/number of exons")

#transcripts with long introns
ggplot(cds_exons_long_transcript, aes(counts.cds/counts.exon))+
  geom_histogram(bins=25, fill="light blue", color="grey")+
  labs(x="# CDS/# exons", title ="Histogram of the ratio number of CDS/number of exons\nin transcripts with long introns")

#transcripts with short introns
ggplot(cds_exons_short_transcript, aes(counts.cds/counts.exon))+
  geom_histogram(bins=25, fill="light blue", color="grey")+
  labs(x="# CDS/# exons", title ="Histogram of the ratio number of CDS/number of exons\nin transcripts with short introns")

```