---
title: "Exploration annotation file"
author: "Emanuela Damieto"
date: "`r Sys.Date()`"
output: html_document
---

```{r setup, include=FALSE}
library("dplyr")
library("ggpubr")
library("ggtree")
library("here")
library("LSD")
library("stringr")
library("tidyr")
```

## Import data
Import the annotation file
```{r import annotation file, echo=FALSE}
#Set the directory and import the data

spruce_pine <- read.table(gzfile(here("reference/gff3/Picab02_codingAll.gff3.gz")), sep = "\t", quote="")
#gff = general feature format

#Add the header
header <- c("seqid", "source", "type", "start", "end", "score", "strand", "phase", "attributes")
colnames(spruce_pine) <- header

#Subset the dataframe to retrieve the information that we want
genes <- subset(spruce_pine, type=="gene")
introns <- subset(spruce_pine, type=="intron")  
cds <- subset(spruce_pine, type=="CDS")
mrna <- subset(spruce_pine, type=="mRNA")

# write_gff3(cds,here("data/gff3/cds.gff3"))
# write_gff3(introns,here("data/gff3/introns.gff3"))
```

## Compare the cumulative intron length with the number of introns
```{r comparison intron length vs # of intorns, echo=FALSE}
#Retrieve the transcript ID (parent)
parent <- sub(".*=","",introns$attributes)
#Look at the transcript with the highest number of introns (74)
#table(parent)[which.max(table(parent))]

#Include transcript ID, length and counts table
introns$parent <- parent
introns$length <- introns$end-introns$start+1
introns$counts <- 1

#create a new dataframe, group by parent
df_introns <- introns %>% group_by(parent)

#summarize the dataframe 
df_introns_cum <- df_introns %>% summarise(
  counts= sum(counts),
  seqid=unique(seqid),
  length=sum(length),
  parent=unique(parent)
)

# plot with x axis not in log scale
comparisonplot(df_introns_cum$counts, log10(df_introns_cum$length), xlab ="number of introns", ylab = "cumulative intron length (log scale)", main="Comparison plot of number of introns vs intron length", cor=T, ylim=c(0,6))

```


```{r median cumulative long vs short, echo=FALSE}
#set the threshold of log intron length to 4 to split in long and short introns
long_introns <- df_introns_cum %>% filter(log10(length)>=4)
short_introns <- df_introns_cum %>% filter(log10(length)<4)
#retrieve the parent ID
long_introns_id <- long_introns$parent
short_introns_id <- short_introns$parent

write.table(short_introns_id, here("data/analysis/short_introns_ID.txt"), col.names=FALSE, quote= FALSE)
write.table(long_introns_id, here("data/analysis/long_introns_ID.txt"), col.names=FALSE, quote= FALSE)

#divide the dataset in long and short introns
df_long_introns_cum <- df_introns_cum %>% filter(parent %in% long_introns_id,)
df_short_introns_cum <- df_introns_cum %>% filter(parent %in% short_introns_id,)

#sum up the counts if they are bigger than 15, group by counts and do the median
median_long_introns <- df_long_introns_cum  %>% mutate(counts=ifelse(counts>15,15,counts)) %>% group_by(counts) %>% summarise(length= median(length))
median_long_introns$type <- "long"

median_short_introns <- df_short_introns_cum  %>% mutate(counts=ifelse(counts>15,15,counts)) %>% group_by(counts) %>% summarise(length= median(length))
median_short_introns$type <- "short"

median <- union_all(median_long_introns, median_short_introns)

ggplot(median, aes(counts, length, color=type))+
  geom_point()+
  geom_smooth(formula = y ~ splines::ns(x, 1))+
  labs(title="Plot of the median of long and short trancripts", x="number of introns", y="Median cumulative length", ylim=c(0,1))+
  theme_classic()

```


## Use another measure of intron length (weighted intron length)
```{r weighted intron length, echo=FALSE}
df_introns_weig <- df_introns %>% summarise(
  counts= sum(counts),
  seqid=unique(seqid),
  length=max(length)/sum(length),
  parent=unique(parent)
)

comparisonplot(df_introns_weig$counts, df_introns_weig$length, xlab ="number of introns", ylab = "weighted intron length", main="Comparison plot of number of introns vs weighted intron length", cor=T, ylim=c(0,1))

#divide the dataset in long and short introns
df_long_introns_weight <- df_introns_weig %>% filter(parent %in% long_introns_id,)
df_short_introns_weight <- df_introns_weig %>% filter(parent %in% short_introns_id,)

#sum up the counts if they are bigger than 15, group by counts and do the median
median_long_introns <- df_long_introns_weight  %>% mutate(counts=ifelse(counts>15,15,counts)) %>% group_by(counts) %>% summarise(length= median(length))
median_long_introns$type <- "long"

median_short_introns <- df_short_introns_weight  %>% mutate(counts=ifelse(counts>15,15,counts)) %>% group_by(counts) %>% summarise(length= median(length))
median_short_introns$type <- "short"

median <- union_all(median_long_introns, median_short_introns)

ggplot(median, aes(counts, length, color=type))+
  geom_point()+
  geom_smooth(formula = y ~ splines::ns(x, 1))+
  labs(title="Plot of the median of long and short trancripts", x="number of introns", y="Median weighted intron length", ylim=c(0,1))+
  theme_classic()


```

## Look at the distribution of intron length and counts along the genome
```{r distribition along the chromosome of intron length, echo=FALSE}
#remove the introns that are not present in chromosomes
df_chr <- df_introns_cum %>% filter(str_detect(seqid,  regex(paste0("^PA_chr[0-9]{2}$"))))

#check the intron length in different chromosomes
ggplot(df_chr, aes(seqid, log10(length)))+
  geom_boxplot()+
  labs(title="Boxplot of cumulative intron length in the chromosomes", x="chr", y="cumulative intron length (log scale)")+
  theme_classic()+
  theme(axis.text.x = element_text(angle = 90))

#check the number of intron in different chromosomes
ggplot(df_chr, aes(seqid, counts))+
  geom_boxplot()+
  labs(title="Boxplot of the number of introns in the chromosomes", x="chr", y="number of introns")+
  theme_classic()+
  theme(axis.text.x = element_text(angle = 90))

```


## Look at the relationship between CDS and intron length 
```{r CDS vs intron length, echo=FALSE}
#Retrieve the transcript ID (parent)
parent_cds <- sub(".*=","",cds$attributes)

#Include transcript ID, length and counts table
cds$parent <- parent_cds
cds$length <- cds$end-cds$start+1
cds$counts <- 1

#create a new dataframe, group by parent
df_cds <- cds %>% group_by(parent)

#summarize the dataframe 
df_cds_cum <- df_cds %>% summarise(
  counts= sum(counts),
  seqid=unique(seqid),
  length=sum(length),
  parent=unique(parent)
)

merge_cds_intron <- df_cds_cum %>% full_join(df_introns_cum, by="parent", suffix=c(".cds",".intron"))

comparisonplot(log10(merge_cds_intron$length.cds), log10(merge_cds_intron$length.intron), xlab="CDS length (log scale)", ylab="Intron length (log scale)", main="Comparisonplot of CDS length vs intron length in log scale")
```

## Compare long and short intron length
Compare the cumulative length of introns in transcripts with long and short introns (or compare the length of introns in transcripts with long and short cumulative intron length)

```{r long vs short introns, echo=FALSE, warning=FALSE}
#comparison intron length in short vs long introns 
mrna <- mrna %>% separate(attributes, c("ID", "Parent", "Others"), ";") %>% mutate (ID= sub("ID=","", ID), Parent= sub("Parent=","",Parent), Others= sub("Others=","",Others))
mrna$mrna_length <- mrna$end-mrna$start+1

mrna_id_length <- mrna %>% select(ID, mrna_length)
short_introns_mrna <- short_introns %>% left_join(mrna_id_length, by= c("parent"="ID"), suffix=c("intron", "mRNA"))
long_introns_mrna <- long_introns %>% left_join(mrna_id_length, by= c("parent"="ID"), suffix=c("intron", "mRNA"))

#Merge the long and short dataset
long_introns_mrna$type <- "long"
short_introns_mrna$type <- "short"

introns_mrna <- bind_rows(long_introns_mrna, short_introns_mrna)

#Use ggpubr package to plot the boxplot
intron_boxplot <- ggboxplot(introns_mrna, x="type", y="length", yscale="log10", color="type", ylim=c(1,1000000), title="Boxplot of log cumulative intron length of transcript\nwith long and short introns", notch=TRUE) 
intron_boxplot + stat_compare_means(label.y.npc = "top", label.x.npc="centre")

#violin plot 
intron_violinplot <- ggviolin(introns_mrna, x="type", y="length", yscale="log10", color="type", ylim=c(1,1000000), title="Boxplot of log cumulative intron length of transcript\nwith long and short introns", fill="type") 
intron_violinplot + stat_compare_means(label.y.npc = "top", label.x.npc="centre")

```

## Compare long and short intron length
Compare the cumulative length of CDS in transcripts with long and short introns (or compare the length of introns in transcripts with long and short cumulative intron length)

```{r CDS long vs short introns, echo=FALSE, warning=FALSE}
#divide the dataset in long and short introns
df_long_cds_cum <- df_cds_cum%>% filter(parent %in% long_introns_id,)
df_short_cds_cum <- df_cds_cum %>% filter(parent %in% short_introns_id,)

#merge CDS and mRNA
long_introns_mrna_cds <- df_long_cds_cum %>% left_join(mrna_id_length, by= c("parent"="ID"), suffix=c("intron", "mRNA"))
short_introns_mrna_cds <- df_short_cds_cum %>% left_join(mrna_id_length, by= c("parent"="ID"), suffix=c("intron", "mRNA"))

#Merge the long and short dataset
long_introns_mrna_cds$type <- "long"
short_introns_mrna_cds$type <- "short"

introns_mrna_cds <- bind_rows(long_introns_mrna_cds, short_introns_mrna_cds)

#Use ggpubr package to plot the boxplot
cds_boxplot <- ggboxplot(introns_mrna_cds, x="type", y="length", yscale="log10", color="type", title="Boxplot of log cumulative CDS length of transcript\nwith long and short introns", notch=TRUE) 
cds_boxplot + stat_compare_means(label.y.npc = "top", label.x.npc="centre")

#violin plot 
cds_violinplot <- ggviolin(introns_mrna_cds, x="type", y="length", yscale="log10", color="type", title="Boxplot of log cumulative CDS length of transcript\nwith long and short introns", fill="type") 
cds_violinplot + stat_compare_means(label.y.npc = "top", label.x.npc="centre")


```



